src/overlays/actors/ovl_En_Mag/z_en_mag.c

```diff
@@ -27,21 +33,553 @@ const ActorInit En_Mag_InitVars = {
+                this->effectPrimColor[1] = 255.0f;
+                this->effectPrimColor[2] = 255.0f;
+                this->effectEnvColor[0] = 255.0f;
+                this->effectEnvColor[1] = 255.0f;
+
+                this->effectFadeInTimer = 40;
+                this->effectFadeInState = 1;
+            }
+        } else if (this->effectFadeInState == 1) {
+            this->effectPrimColor[0] += -2.125f;
+            this->effectEnvColor[0] += -1.375f;
+
+            this->effectPrimLodFrac += 2.4f;
+
+            this->effectFadeInTimer--;
+
+            if (this->effectFadeInTimer == 0) {
+                this->effectPrimLodFrac = 128.0f;
+
+                this->effectPrimColor[0] = 170.0f;
+                this->effectEnvColor[0] = 200.0f;
+
+                this->effectFadeInTimer = 32;
+                this->effectFadeInState = 2;
+            }
+        }
+
+        if (this->effectAlpha > 160) {
+            this->mainAlpha += VREG(5);
+            if (this->mainAlpha >= 210.0f) {
+                this->mainAlpha = 210.0f;
+            }
+
+            if (this->mainAlpha >= 210) {
+                this->subAlpha += VREG(6);
+                if (this->subAlpha >= 255.0f) {
+                    this->subAlpha = 255.0f;
+                }
+
+                if (this->subAlpha >= 200) {
+                    this->copyrightAlpha += this->copyrightAlphaStep;
+                    if (this->copyrightAlpha >= 255.0f) {
+                        this->copyrightAlpha = 255.0f;
+                        this->globalState = MAG_STATE_DISPLAY;
+                        sDelayTimer = 20;
+                    }
+                }
+            }
+        }
+    } else if (this->globalState == MAG_STATE_FADE_OUT) {
+        this->effectAlpha -= this->fadeOutAlphaStep;
+        if (this->effectAlpha < 0.0f) {
+            this->effectAlpha = 0.0f;
+        }
+
+        this->mainAlpha -= this->fadeOutAlphaStep;
+        if (this->mainAlpha < 0.0f) {
+            this->mainAlpha = 0.0f;
+        }
+
+        this->subAlpha -= this->fadeOutAlphaStep;
+        if (this->subAlpha < 0.0f) {
+            this->subAlpha = 0.0f;
+        }
+
+        this->copyrightAlpha -= this->copyrightAlphaStep;
+        if (this->copyrightAlpha < 0.0f) {
+            this->copyrightAlpha = 0.0f;
+            this->globalState = MAG_STATE_POST_DISPLAY;
+        }
+    }
+
+    if (this->globalState == MAG_STATE_INITIAL) {
+        if (Flags_GetEnv(globalCtx, 3)) {
+            this->effectFadeInTimer = 40;
+            this->globalState = MAG_STATE_FADE_IN;
+        }
+    } else if (this->globalState == MAG_STATE_DISPLAY) {
+        if (Flags_GetEnv(globalCtx, 4)) {
+            this->globalState = MAG_STATE_FADE_OUT;
+        }
+    }
+}
+
+void EnMag_DrawTextureI8(Gfx** gfxp, void* texture, s16 texWidth, s16 texHeight, s16 rectLeft, s16 rectTop,
+                         s16 rectWidth, s16 rectHeight, u16 dsdx, u16 dtdy) {
```

Comment:
I think the distinction between image and texture is important. `EnMag_DrawTextureI8` draws a single I8 texture while `EnMag_DrawImageRGBA32` draws as many textures as it needs to do the whole image. So imo the functions and arguments shouldn't have consistent names since they kinda mean different things.

The other argument names (`rectXXX` and `dsdx`/`dtdy`) are mostly based on what I did in other code files like `z_parameter.c`. I don't know if they are necessarily the best but I feel like this is something that will make a lot more sense to change once we are used to all the different ways/patterns used to draw textures.

Reply:
I guess it makes sense if you see "image" as a block of data and "texture" as the thing thats actually drawn onto the rectangle/triangle, but the image (what you call texture I think) is only part of the data that is used to compute what is effectively drawn onto the triangle/rectangle with respects to perspective and stuff, and what I call texture.
I don't think there's a very clear definition of "texture", but I'm not sure what term would be used for what the combiner outputs if not "texture" then (with inputs being image/color registers/shading...).